### DISCUSSION TITLE
Learning Rate schedules (ReduceLROnPlateau, cosine) support

### DISCUSSION THREAD

**Discussion by kirillbobyrev** [huggingface/candle#2224] [General]
Most Deep Learning frameworks support Learning Rate reduction schedulers (cosine, ReduceLROnPLateau etc), fixed schedule, etc. While not very hard to implement manually, out-of-the-box support would definitely be very nice.

Single schedules are not too complicated, but joint schedules and some extra support is definitely useful.

E.g. [PyTorch ReduceLROnPlateau](https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.ReduceLROnPlateau.html) or Jax's [Optim schedules](https://optax.readthedocs.io/en/latest/api/optimizer_schedules.html).

This is the first time I'm seeing and reading Candle's code, so please let me know if there is better channel/place to ask questions like this.

**Comment by EricLBuehler** [comment]
@kirillbobyrev, I opened #2225 which refactors the structure to allow this to be implemented.

**Reply by kirillbobyrev** [reply]
Thanks a lot!

**Comment by brylee10** [comment]
I also found a need for this feature, so I made a related PR here: https://github.com/huggingface/candle/pull/2861
